feat: introduce api leveling proposal (#3317)
meta-llama/llama-stack: docs/source/apis/api_leveling.md

# Llama Stack API Stability Leveling

In order to provide a stable experience in Llama Stack, the various APIs need different stability levels indicating the level of support, backwards compatability, and overall production readiness.

## Different Levels

> ### v1alpha1

Reviewer: `/alpha/v1` or `/beta/v1/` feel slightly more readable?

Author: I don't think `v1alpha2` would exist. If it helps clear it up I can switch both to `v1alpha` and `v1beta` without the trailing 1. I agree, It could be nice to extend this in the case of a `v2` api.

> - little/no expectation of support between versions, breaking changes are permitted
> - datatypes and parameters can break, routes can be added and removed.

Reviewer 2: ```suggestion
- Little to no expectation of support between versions
- Breaking changes are permitted
- Datatypes and parameters can break
- Routes can be added and removed
```

### v1beta1

> - API routes ensured between versions, but parameters and return types are not. However, the changes should not be drastic.
> - API, besides minor fixes and adjustments, should be _almost_ v1.

Reviewer 2: ```suggestion
- API routes ensured between versions
- Parameters and return types are not
```

> ### v1 (stable)
> 
> - backwards compatible between z-streams, y-stream breaking changes must go through the proper approval and announcement process.
> - datatypes for a route, and its return types cannot change between z-streams. y-stream datatype changes should be sparing, unless the changes are additional net-new parameters.
> - must have proper conformance testing as outlined in https://github.com/llamastack/llama-stack/issues/3237

Reviewer 2: ```suggestion
### v1

- Considered stable
- Backwards compatible between Z-streams
  - Y-stream breaking changes must go through the proper approval and announcement process.
- Datatypes for a route and its return types cannot changed between Z-streams
  - Y-stream datatype changes should be sparing, unless the changes are additional net-new parameters
- Must have proper conformance testing as outlined in https://github.com/llamastack/llama-stack/issues/3237
```

> - must have proper conformance testing as outlined in https://github.com/llamastack/llama-stack/issues/3237

Reviewer 3: ```suggestion
- Must have proper conformance testing as outlined in [issue #3237](https://github.com/llamastack/llama-stack/issues/3237)
```

### Approval and Announcement Process for Breaking Changes

- **PR Labeling**: Any pull request that introduces a breaking API change must be clearly labeled with `breaking-change`.
- **Changelog Entry**: The PR must also update the changelog to describe the change, its impact, and any migration steps.
- **Maintainer Review**: At least one maintainer must explicitly acknowledge the breaking change during review.
- **Announcement**: Breaking changes require inclusion in release notes and, if applicable, a separate communication (e.g., mailing list, Slack, or GitHub Discussions) prior to release.

## Enforcement

> ### Migration of API routes under /v1alpha1, /v1beta1, and /v1

Reviewer 3: ```suggestion
### Migration of API routes under `/v1alpha1`, `/v1beta1`, and `/v1`
```

> Instead of placing every API under `v1`, any API that is not fully stable or complete should go under `v1alpha1` or `v1beta1`. For example, post_training belongs here, as well as any OpenAI-compatible API whose surface does not exactly match the upstream OpenAI API it mimics.

Reviewer 2: ```suggestion
Instead of placing every API under `v1`, any API that is not fully stable or complete should go under `v1alpha1` or `v1beta1`. For example, at the time of this writing `post_training` belongs here, as well as any OpenAI-compatible API whose surface does not exactly match the upstream OpenAI API it mimics.
```

> Instead of placing every API under `v1`, any API that is not fully stable or complete should go under `v1alpha1` or `v1beta1`. For example, post_training belongs here, as well as any OpenAI-compatible API whose surface does not exactly match the upstream OpenAI API it mimics.

Reviewer 3: ```suggestion
Instead of placing every API under `/v1`, any API that is not fully stable or complete should go under `/v1alpha1` or `/v1beta1`. For example, at the time of this writing , `post_training` belongs here, as well as any OpenAI-compatible API whose surface does not exactly match the upstream OpenAI API it mimics.
```

> This migration is cricual as we get Llama Stack in the hands of users who intend to productize various APIs. A clear view of what is stable and what is actively being developed will enable customers to pick and choose various APIs to build their products on.

Reviewer 2: ```suggestion
This migration is crucial as we get Llama Stack in the hands of users who intend to productize various APIs. A clear view of what is stable and what is actively being developed will enable users to pick and choose various APIs to build their products on.
```

This migration will be a breaking change for any api moving out of `/v1`. Ideally, this should happen before 0.3.0 and especially 1.0.0.

> ### x-stability tags in the openapi spec for oasdiff

Reviewer 3: ```suggestion
### `x-stability` tags in the OpenAPI spec for oasdiff
```

> x-stability tags allow for tools like oasdiff to enforce different rules for different stability levels, these tags should match the routes: https://github.com/oasdiff/oasdiff/blob/main/docs/STABILITY.md

Reviewer 3: ```suggestion
`x-stability` tags allow tools like oasdiff to enforce different rules for different stability levels; these tags should match the routes: [oasdiff stability](https://github.com/oasdiff/oasdiff/blob/main/docs/STABILITY.md)
```

### Testing

> The testing of each stable API is already outlined in https://github.com/llamastack/llama-stack/issues/3237 and is being worked on. These sorts of conformance tests should apply primarily to `/v1` APIs only, with `v1alpha` and `v1beta1` having any tests the maintainers see fit as well as basic testing to ensure the routing works properly.

Reviewer 3: ```suggestion
The testing of each stable API is already outlined in [issue #3237](https://github.com/llamastack/llama-stack/issues/3237) and is being worked on. These sorts of conformance tests should apply primarily to `/v1` APIs only, with `/v1alpha1` and `/v1beta1` having any tests the maintainers see fit, as well as basic testing to ensure the routing works properly.
```

# Llama Stack API Stability Leveling

In order to provide a stable experience in Llama Stack, the various APIs need different stability levels indicating the level of support, backwards compatability, and overall production readiness.

## Different Levels

### v1alpha

- Little to no expectation of support between versions
- Breaking changes are permitted
- Datatypes and parameters can break
- Routes can be added and removed

### v1beta

- API routes ensured between versions
- Parameters and return types are not
- API, besides minor fixes and adjustments, should be _almost_ v1. Changes should not be drastic.

### v1 (stable)

> - Considered stable

Reviewer 4: I think this is key: what are the criteria to graduate APIs from alpha to beta to stable?

Reviewer 4: Thanks for adding those!

- Backwards compatible between Z-streams
  - Y-stream breaking changes must go through the proper approval and announcement process.
- Datatypes for a route and its return types cannot changed between Z-streams
  - Y-stream datatype changes should be sparing, unless the changes are additional net-new parameters
- Must have proper conformance testing as outlined in https://github.com/llamastack/llama-stack/issues/3237

### Approval and Announcement Process for Breaking Changes

> - **PR Labeling**: Any pull request that introduces a breaking API change must be clearly labeled with `breaking-change`.

Reviewer 5: is this for v1alpha only? Other stability levels do not seem to allow breaking-change in any case, in which case do we need this process?

> - **PR Labeling**: Any pull request that introduces a breaking API change must be clearly labeled with `breaking-change`.

Reviewer 4: I'm wondering how useful this will be in practice, but I guess it doesn't hurt.

Author: its really just another mechanism to ensure these types of changes are purposeful and not accidental. I think applying a bit of friction here is a good thing.

Reviewer 4: came back to this, I don't think this is useful especially if we have a proper CHANGELOG but won't block for this, will see if this happens in practice :)

> - **Changelog Entry**: The PR must also update the changelog to describe the change, its impact, and any migration steps.

Reviewer 4: That's a dup of the "Announcement" section where we include release note already.

Reviewer 4: Announcement through the CHANGELOG. We don't need to separate entries. We can remove the CHANGELOG piece and indicate in the Announcement that the CHANGELOG must reflect the breaking change in a dedicated section.

Reviewer 4: Still not addressed, also the CHANGELOG is typically updated async today so this won't quite work unless we change that process.

> - **Maintainer Review**: At least one maintainer must explicitly acknowledge the breaking change during review.

Reviewer 4: If a maintainer approves a PR that includes a breaking change, what's left to do as an "explicite acknowledgement"?

> - **Announcement**: Breaking changes require inclusion in release notes and, if applicable, a separate communication (e.g., mailing list, Slack, or GitHub Discussions) prior to release.

Reviewer 4: We don't have a ML and we don't use Slack. Perhaps replace with "Chat"?

Author: true, I was putting these here for future-proofing, but I can re-word!

## Enforcement

### Migration of API routes under `/v1alpha`, `/v1beta`, and `/v1`

Instead of placing every API under `/v1`, any API that is not fully stable or complete should go under `/v1alpha` or `/v1beta`. For example, at the time of this writing,  `post_training` belongs here, as well as any OpenAI-compatible API whose surface does not exactly match the upstream OpenAI API it mimics.

> This migration is crucual as we get Llama Stack in the hands of users who intend to productize various APIs. A clear view of what is stable and what is actively being developed will enable users to pick and choose various APIs to build their products on.

Reviewer 4: ```suggestion
This migration is crucial as we get Llama Stack in the hands of users who intend to productize various APIs. A clear view of what is stable and what is actively being developed will enable users to pick and choose various APIs to build their products on.
```

Author: thanks!

# Llama Stack API Stability Leveling

In order to provide a stable experience in Llama Stack, the various APIs need different stability levels indicating the level of support, backwards compatability, and overall production readiness.

## Different Levels

### v1alpha

- Little to no expectation of support between versions
- Breaking changes are permitted
- Datatypes and parameters can break
- Routes can be added and removed

#### Graduation Criteria

- an API can graduate from `v1alpha` to `v1beta` if the team has identified the minimum set of capabilities every provider must implement. Also CRUD must stay stable once in `v1beta`.

> - for OpenAI APIs, a comparison to the OpenAI spec for the specific API can be done to ensure completeness.

Reviewer 6: probably this needs to be linked to a specific OpenAI API version 😅

### v1beta

- API routes ensured between versions

> - Parameters and return types are not

Reviewer 6: is this sentence incomplete?

- API, besides minor fixes and adjustments, should be _almost_ v1. Changes should not be drastic.

#### Graduation Criteria

- an API can graduate from `v1beta` to `v1` if the API surface and datatypes are complete as identified by the team. The parameters and return types that are mandatory for each route are stable. All aspects of graduating from `v1alpha1` to `v1beta` apply as well.
- Optional parameters, routes, or parts of the return type can be added after graduating to `v1`

### v1 (stable)

- Considered stable

> - Backwards compatible between Z-streams

Author: I think I had this before and was removed for simplicity during review process.

- Datatypes for a route and its return types cannot changed between Z-streams
  - Y-stream datatype changes should be sparing, unless the changes are additional net-new parameters
- Must have proper conformance testing as outlined in https://github.com/llamastack/llama-stack/issues/3237

### API Stability vs. Provider Stability

The leveling introduced in this document relates to the stability of the API and not specifically the providers within the API.

Providers can iterate as much as they want on functionality as long as they work within the bounds of an API. If they need to change the API, then the API should not be `/v1`, or those breaking changes can only happen on a y-stream release basis.

### Approval and Announcement Process for Breaking Changes

- **PR Labeling**: Any pull request that introduces a breaking API change must be clearly labeled with `breaking-change`.
- **Maintainer Review**: At least one maintainer must explicitly acknowledge the breaking change during review by applying the `breaking-change` label. An approval must come with this label or the acknowledgement this label has already been applied.

> - **Announcement**: Breaking changes require inclusion in release notes and, if applicable, a separate communication (e.g., Discord, Github Issues, or GitHub Discussions) prior to release.

Reviewer 6: This piece is important.

This migration is crucial as we get Llama Stack in the hands of users who intend to productize various APIs. A clear view of what is stable and what is actively being developed will enable users to pick and choose various APIs to build their products on.

This migration will be a breaking change for any API moving out of `/v1`. Ideally, this should happen before 0.3.0 and especially 1.0.0.

### `x-stability` tags in the OpenAPI spec for oasdiff

`x-stability` tags allow tools like oasdiff to enforce different rules for different stability levels; these tags should match the routes: [oasdiff stability](https://github.com/oasdiff/oasdiff/blob/main/docs/STABILITY.md)

### Testing

The testing of each stable API is already outlined in [issue #3237](https://github.com/llamastack/llama-stack/issues/3237) and is being worked on. These sorts of conformance tests should apply primarily to `/v1` APIs only, with `/v1alpha` and `/v1beta` having any tests the maintainers see fit as well as basic testing to ensure the routing works properly.

> ## Next Steps

Reviewer 6: it feels like this section should go in the PR description, not in this doc?

# Llama Stack API Stability Leveling

In order to provide a stable experience in Llama Stack, the various APIs need different stability levels indicating the level of support, backwards compatability, and overall production readiness.

## Different Levels

### v1alpha

- Little to no expectation of support between versions
- Breaking changes are permitted
- Datatypes and parameters can break
- Routes can be added and removed

#### Graduation Criteria

> - an API can graduate from `v1alpha` to `v1beta` if the API surface is complete. Meaning no net-new routes will be added that all providers must implement.

Reviewer 4: ```suggestion
- an API can graduate from `v1alpha` to `v1beta` if the team has identified the minimum set of capabilities every provider must implement. Also CRUD must stay stable once in beta.
```

I think the "no net-new" is a bit unrealistic and more aspirational than enforceable; most APIs don’t stop growing once they hit beta or GA.  But optional new routes/extensions are fine.

Author: yeah this is fair, will change

- for OpenAI APIs, a comparison to the OpenAI spec for the specific API can be done to ensure completeness.

### v1beta

- API routes ensured between versions
- Parameters and return types are not
- API, besides minor fixes and adjustments, should be _almost_ v1. Changes should not be drastic.

#### Graduation Criteria

> - an API can graduate from `v1beta` to `v1` if the API surface and datatypes are complete. Meaning no net-new routes will be added and the parameters and return types that are mandatory for each route are stable.

Reviewer 4: ditto on no net-new.

- Optional parameters, routes, or parts of the return type can be added after graduating to `v1`

### v1 (stable)

- Considered stable
- Backwards compatible between Z-streams
  - Y-stream breaking changes must go through the proper approval and announcement process.
- Datatypes for a route and its return types cannot change between Z-streams
  - Y-stream datatype changes should be sparing, unless the changes are additional net-new parameters
- Must have proper conformance testing as outlined in https://github.com/llamastack/llama-stack/issues/3237

### API Stability vs. Provider Stability

The leveling introduced in this document relates to the stability of the API and not specifically the providers within the API.

Providers can iterate as much as they want on functionality as long as they work within the bounds of an API. If they need to change the API, then the API should not be `/v1`, or those breaking changes can only happen on a y-stream release basis.

> ### Approval and Announcement Process for Breaking Changes

Reviewer 7: should probably also include something like this to define a protocol for when there is a breaking change - https://github.com/llamastack/llama-stack/pull/3260. A PR that is titled a specific way will not fail the oasdiff check.

Author: so by this you mean: I should add a bullet here describing how the PR title and commit message should include an indicator of a breaking change? I can add that!

Author: I added a section here, luckily conventional commits outlines how to handle this: https://www.conventionalcommits.org/en/v1.0.0/#specification
